### ISSUE TITLE
Receiver Type Inference

### ISSUE THREAD

**Issue by chloestefantsova** [dart-lang/language#3527]
This issue is created in response to https://github.com/dart-lang/language/issues/3523.

Consider the following program.

```dart
void main() {
  List<num> xs = [1].toList();
  xs.add(1.5); // Throws.
}
```

Currently the type inference mechanism in Dart always infers receivers of all kinds of invocations in the empty type context. So, the type literal `[1]` is inferred in the empty context, ultimately determining the type argument of the list literal from the types of the elements rather than from the type of the variable `xs` being initialized.

The current feature proposal suggests using type contexts of structural kind called _context type terms_ that are suitable for encoding of the downwards type information for the receivers. In the example above, the structural context for `[1]` as the receiver of the `toList` method invocation can be described as `{toList(): (...) → List<num>}`. That structure indicates that `[1]` is supposed to have a member called `toList` that can be invoked and returns a value of type `List<num>` upon the invocation. The type information encoded this way can help with inferring the missing type arguments from the types hinted by the surrounding type context. To enable passing the new kind of type information between the inference phases, the current proposal suggests considering context type terms first-class type schemas in the type inferring mechanism. 

As a related concern, the current feature proposal suggests replacing the currently used eager approach for deciding on the inferred type with a delayed approach, where the type information is gathered from all available sources, and the most appropriate solution to the constraint system is picked among the collected ones. The delayed approach allows combining the collected type information too, for example, when some of it comes in the form of context type terms during the downwards inference phase, and some other type information comes in the form of nominal types during the upwards inference phase.

**Comment by AbdeMohlbi** [comment]
Hello i just wanted to confirm is this related to the issue with this code ?
This code gives an error in the anlyzer
```dart
import 'dart:math';

int medianSum(int a, int b, int c) {
  return a + b + c - min(a, min(b, c)) - max(a, max(b, c));
}
```
Note if u replace +/- with ^ the code would compile and the analyzer won't complain about it 
Cc @FMorschel

**Comment by chloestefantsova** [comment]
> Hello i just wanted to confirm is this related to the issue with this code ?

I think this specific compile-time error is unrelated to Receiver Type Inference, and rather has to do with the way `operator -` is defined. The formal parameter is always `num`, regardless of the receiver or the surrounding context, which causes the `T` in the invocation of `max` to be inferred as `num`, further propagating `num` to be the static type of the entire return expression.

/cc @eernstg

**Comment by eernstg** [comment]
There's even one more layer of complexity when it comes to numeric types: https://github.com/dart-lang/language/blob/main/accepted/2.12/nnbd/number-operation-typing.md.

First note that the additive operators (including `+` and `-`) associate to the left. This implies that the receiver of the first `-` operator (that is `a + b + c`) has static type `int`. The context type of the entire expression is `int` (obtained from the return type of the function). However, the context type for the receiver of the second `-` operator is then empty (because it's a receiver), and this implies that the context type for `min(a, min(b, c))` is `num`. Hence, the inferred type argument is also `num`, and so is the type of the expression `min(a, min(b, c))`. This implies that the receiver for the last `-` operator has type `num`, which causes the failure.

@chloestefantsova, with your receiver type inference proposal, wouldn't this change the crucial context type such that the first `min` would have the inferred type argument `int` and the overall typing would succeed?

**Comment by chloestefantsova** [comment]
> [@chloestefantsova](https://github.com/chloestefantsova), with your receiver type inference proposal, wouldn't this change the crucial context type such that the first `min` would have the inferred type argument `int` and the overall typing would succeed?

My understanding is that it would still be `num`. The formal parameter to `operator -` is `num`, regardless of the receiver, so we would infer `num` as the type argument for `max`, which would make `num` the static type of the invocation of `max`. Then, since the type of the actual argument of `operator -` isn't `int`, but `num`, the return type of `operator -` would be `num`, and similarly for the `operator -`s and `operator +`s earlier in the expression. So, the overall static type of the entire return expression would still be `num`.

My guess is that it's the special casing of `operator -` and `operator +` that stands in the way in this specific example. The return type of those operators depend on the static type of the actual parameter.

**Comment by FMorschel** [comment]
> The formal parameter to `operator -` is `num`, regardless of the receiver. 

This is not entirely true.

> spec (section 17.31) reads:
> The static type of an additive expression is usually determined by the signature given in the declaration of the operator used. However, invocations of the operators + and - of class int, double and num are treated specially by the typechecker.

This changes the parameter to `int` in cases like:

```dart
int x = 0;
x += foo;
``` 

So I suspect for the case above it would eventually work if this issue were complete.

**Comment by AbdeMohlbi** [comment]
> There's even one more layer of complexity when it comes to numeric types: https://github.com/dart-lang/language/blob/main/accepted/2.12/nnbd/number-operation-typing.md.
> 
> First note that the additive operators (including `+` and `-`) associate to the left. This implies that the receiver of the first `-` operator (that is `a + b + c`) has static type `int`. The context type of the entire expression is `int` (obtained from the return type of the function). However, the context type for the receiver of the second `-` operator is then empty (because it's a receiver), and this implies that the context type for `min(a, min(b, c))` is `num`. Hence, the inferred type argument is also `num`, and so is the type of the expression `min(a, min(b, c))`. This implies that the receiver for the last `-` operator has type `num`, which causes the failure.
> 
> @chloestefantsova, with your receiver type inference proposal, wouldn't this change the crucial context type such that the first `min` would have the inferred type argument `int` and the overall typing would succeed?

But why does breaking the function into multiple variables seems to work as expected then ?

**Comment by eernstg** [comment]
If we only have this `return a + b + c - min(a, min(b, c));` then the receiver `a + b + c` will have type `int` and the entire expression will have context type `int`, and in this case the context type for the operand `min(a, min(b, c))` [will be](https://github.com/dart-lang/language/blob/main/accepted/2.12/nnbd/number-operation-typing.md#improved-context-type) `int`, too.

**Comment by chloestefantsova** [comment]
> > The formal parameter to `operator -` is `num`, regardless of the receiver.
> 
> This is not entirely true.
> 
> > spec (section 17.31) reads:

Right. I forgot about that piece of special treatment of math operators. Then, I believe, Receiver Type Inference would infer `int` for the given example.

**Comment by AbdeMohlbi** [comment]
> If we only have this `return a + b + c - min(a, min(b, c));` then the receiver `a + b + c` will have type `int` and the entire expression will have context type `int`, and in this case the context type for the operand `min(a, min(b, c))` [will be](https://github.com/dart-lang/language/blob/main/accepted/2.12/nnbd/number-operation-typing.md#improved-context-type) `int`, too.

I think i see now so we are under the third case here :
```readme
If int <: C, not num <: C, and T <: int, then the context type of e2 is int.
If double <: C, not num <: C, and not T <: double, then the context type of e2 is double.
Otherwise, the context type of e2 is num
```
